benchmarking/daint_single_node/run_benchmark.py

```diff
@@ -0,0 +1,183 @@
+# script is generated and submitted. Optionally, we wait for the completion of the SLURM
+# job.
+
+# 2021/01/22 Oliver Fuhrer, Vulcan Inc, oliverf@vulcan.com
+
+import math
+import os
+import sys
+import shutil
+import hashlib
+import yaml
+import click
+import fv3config
+
+# constants
+DAINT_SECTION = 'hybrid' # one of hybrid or multicore
+FV3_NUMBER_OF_TILES = 6
+
+# Piz Daint specification
+if DAINT_SECTION == 'hybrid':
+    SOCKETS_PER_NODE = 1
+    CORES_PER_SOCKET = 12
+    HAS_HYPERTHREADING = True
+if DAINT_SECTION == 'multicore':
+    SOCKETS_PER_NODE = 2
+    CORES_PER_SOCKET = 18
+    HAS_HYPERTHREADING = True
+
+@click.command()
+@click.option('--blocksize', default=-1, help='block size to use for physics')
+@click.option('--executable', default='./fv3.exe', help='path to the FV3GFS executable')
+@click.option('--force/--no-force', default=False, help='overwrite existing run directory (if it exists)')
+@click.option('--hyperthreading/--no-hyperthreading', default=False, help='switch hyperthreading on or off')
+@click.option('--module_env', default=None, help='file to source to pre-load modules')
+@click.option('--nodes_per_tile_side', default=1, help='number of compute nodes per tile side')
+@click.option('--partition', default='normal', help='parittion to submit job to')
+@click.option('--rank_layout', default=-1, help='Choose which rank layout to use (1, 2, 3, ...)')
+@click.option('--threads_per_rank', default=1, help='number of OpenMP threads per rank')
+@click.option('--timesteps', default=61, help='number of timesteps')
+@click.option('--wait/--no-wait', default=True, help='wait for SLURM job to complete')
+@click.argument('yaml_file', type=click.File('r'), nargs=1)
+@click.argument('run_directory', type=str, nargs=1)
+def run_benchmark(nodes_per_tile_side, threads_per_rank, hyperthreading, executable, 
+    rank_layout, blocksize, partition, force, timesteps, wait, module_env, yaml_file, run_directory):
+
+    # echo command
+    print('\nRunning command:')
+    print('', ' '.join(sys.argv))
+
+    # derived variables for Piz Daint
+    if hyperthreading:
+        assert HAS_HYPERTHREADING
+        threads_per_core = 2
+    else:
+        threads_per_core = 1
+    max_cores_per_node = SOCKETS_PER_NODE * CORES_PER_SOCKET
+    max_threads_per_node = max_cores_per_node * threads_per_core
+
+    # define SLURM parameters
+    slurm_nodes = FV3_NUMBER_OF_TILES * nodes_per_tile_side**2
+    slurm_ntasks_per_node = max_threads_per_node // threads_per_rank
+    assert threads_per_rank * slurm_ntasks_per_node == max_threads_per_node
+    slurm_ntasks_per_core = threads_per_core
+    slurm_cpus_per_task = threads_per_rank
+    if hyperthreading:
+        slurm_hint="multithread"
+    else:
+        slurm_hint="nomultithread"
+    if DAINT_SECTION == 'hybrid':
+        slurm_constraint = 'gpu'
+    if DAINT_SECTION == 'multicore':
+        slurm_constraint = 'mc'
+
+    # SLURM job
+    slurm_job = f"""#!/bin/bash -l
```

Comment:
what is the reason behind creating the slurm file here?

to me it would feel more intuitive doing this at least after MPI configuration and benchmark parameters, potentially even just before launching the job

Reply:
Agreed. Was mostly because some of the `slurm_`* variables were being used. But I moved the creation of the SLURM job down. Makes more sense.